crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -10793,6 +10807,26 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                 )
             })),
 
+            (Type::NewTypeInstance(newtype), right) => Some(
+                visitor.visit((left, op, right), || { self.infer_binary_type_comparison(
+                    newtype.concrete_base_type(self.db()),
+                    op,
+                    right,
+                    range,
+                    visitor,
+                )
+            })),
+
+            (left, Type::NewTypeInstance(newtype)) => Some(
+                visitor.visit((left, op, right), || { self.infer_binary_type_comparison(
+                    left,
+                    op,
+                    newtype.concrete_base_type(self.db()),
+                    range,
+                    visitor,
+                )
+            })),
```

Comment:
I still don't see why these special cases are needed. It seems like `infer_rich_type_comparison` is failing to find matching bindings, but these bindings in typeshed should be fine:

```py
    def __lt__(self, value: float, /) -> bool: ...
    def __le__(self, value: float, /) -> bool: ...
    def __gt__(self, value: float, /) -> bool: ...
    def __ge__(self, value: float, /) -> bool: ...
```

`float` there in type position should(?) expand to `float | int`, and a `NewType` of `float` should be assignable to that...

Reply:
Ok progress: This special case is needed because it allows the recursive call to reach the `Type::Union` special case above. Without that, the generic `try_dunder` -> `infer_rich_comparison` -> `call_dunder` path ends up asking whether the `NewType` is assignable to the parameters of e.g. `float.__lt__`. It turns out it is assignable to the positional parameter, which benefits from the `float` -> `int | float` transformation, but it's _not_ assignable to the `self` parameter, which is just `float`.